Sources/XMLCoder/Box/Box.swift

```diff
@@ -0,0 +1,238 @@
+        guard case .dictionary(let box) = self else {
+            return nil
+        }
+        return box
+    }
+    
+    var xmlString: String {
+        switch self {
+        case .null(let box): return box.xmlString
+        case .bool(let box): return box.xmlString
+        case .decimal(let box): return box.xmlString
+        case .signedInteger(let box): return box.xmlString
+        case .unsignedInteger(let box): return box.xmlString
+        case .floatingPoint(let box): return box.xmlString
+        case .string(let box): return box.xmlString
+        case .array(let box): return box.xmlString
+        case .dictionary(let box): return box.xmlString
+        }
+    }
+    
+    init() {
+        self = .null(.init())
+    }
+    
+    init(_ unboxed: Bool) {
+        self = .bool(.init(unboxed))
+    }
+    
+    init(_ unboxed: Decimal) {
+        self = .decimal(.init(unboxed))
+    }
+    
+    init<Integer: SignedInteger>(_ unboxed: Integer) {
+        self = .signedInteger(.init(unboxed))
+    }
+    
+    init<Integer: UnsignedInteger>(_ unboxed: Integer) {
+        self = .unsignedInteger(.init(unboxed))
+    }
+    
+    init<Float: BinaryFloatingPoint>(_ unboxed: Float) {
+        self = .floatingPoint(.init(unboxed))
+    }
+    
+    init(_ unboxed: String) {
+        self = .string(.init(unboxed))
+    }
+    
+    init(_ unboxed: [Box]) {
+        self = .array(.init(unboxed))
+    }
+    
+    init(_ unboxed: [String: Box]) {
+        self = .dictionary(.init(unboxed))
+    }
+    
+    internal func unbox() throws -> Bool? {
+        switch self {
+        case .bool(let box): return box.unbox()
+        case _: return nil
+        }
+    }
+    
+    internal func unbox() throws -> Decimal? {
+        switch self {
+        case .decimal(let box): return box.unbox()
+        case _: return nil
+        }
+    }
+    
+    internal func unbox<Integer: BinaryInteger>() throws -> Integer? {
+        switch self {
+        case .signedInteger(let box): return box.unbox()
+        case .unsignedInteger(let box): return box.unbox()
+        case _: return nil
+        }
+    }
+    
+    internal func unbox<Float: BinaryFloatingPoint & FloatingPoint>() throws -> Float? {
+        switch self {
+        case .floatingPoint(let box): return box.unbox()
+        case _: return nil
+        }
+    }
+    
+    internal func unbox() throws -> String? {
+        switch self {
+        case .string(let box): return box.unbox()
+        case _: return nil
+        }
+    }
+    
+    internal func unbox() throws -> [Box]? {
+        switch self {
+        case .array(let box): return box.unbox()
+        case _: return nil
+        }
+    }
+    
+    internal func unbox() throws -> [String: Box]? {
+        switch self {
+        case .dictionary(let box): return box.unbox()
+        case _: return nil
+        }
+    }
+}
+
+extension Box: Equatable {
+    static func == (lhs: Box, rhs: Box) -> Bool {
```

Comment:
hm, wouldn't `Equatable` implementation for `Box` derive automatically with the correct implementation? My thinking was that you can get a compiler-generated `Equatable` for all enums with associated values that conform to `Equatable`, which I think those associated values should have anyway

Reply:
Fixed.

Reply:
Having it all in my head at once, I think if you require `Equatable` for all `Box` implementors you could avoid this custom implementation too. Like this:

```
protocol Box: Equatable {
   var xmlString: String { get }
   var isFragment: Bool { get }
}

extension Array: Box {
//...
}

extension Bool: Box {
//...
}
```

If you use standard `Array` instead of new `ArrayBox`, you'd have `Equatable` implementations on these wrapped types already.

Reply:
> If you use standard Array instead of new ArrayBox, you'd have Equatable implementations on these wrapped types already.

One of my main driving reasons for using Box types is to encapsulate the behavior. I want to be able to change the internal behavior without having to edit dozens of lines of code sprinkled all over the project.

Case in point: Changing `DictionaryBox` to have `OrderedDictionary` semantic, next. Which is my main motivation for this PR.

---

I'm also think that the whole logic within all those `func box`/`func unbox` in `_XMLEncoder`/`_XMLDecoder` should be moved into their respective individual Box types to make things easier to test and more organized/encapsulated.

Right now `XMLEncoder.swift`/`XMLDecoder.swift` are 600/800loc monsters (even after I moved a whole bunch of code out there already!) that will be a pain to maintain in the future.